### ISSUE TITLE
switch OO scheme

### ISSUE THREAD

**Issue by dankelley** [dankelley/mooring#5]
The present scheme builds up moorings by defining a variant of `+` that works on S3 objects. The problem is that the first object in a list `a+b+c` will have some class, and so will the second.  Therefore, the sum must have the same, for the operation to commute.

This is done presently by using a list for the overall mooring, and another list (of length 1) for each component.  This is awkward for the user, who must do e.g.
```R
f <- float()
f[[1]]$CD
```
to find the $C_D$ of the float object.  I don't think users will guess that, and it's a lot of typing.  Basically, it's confusing to have a list of lists (although that's how matlab objects often get created in R, e.g. the rafos float data that CL used during thesis work).

That's problem 1 with the present scheme.  The other is that this `+` operator is not very discoverable based on its name.  And, even then, I think q lot of users won't know to type
```R
?`+.mooring`
```
to get its documentation.

I am going to switch this.  The new scheme will use a function `mooring()` that takes objects of certain allowed subclass (e.g. a wire element would inherit from the two classes `wire` and `mooring`) and produce something of another class (maybe inheriting from `mooring` only ... I've not decided yet).  This eliminates the need for a list of lists.

As a short-term measure, for a few days anyway, I'll make the `+` operation spit out a message as to the new way.

**Plan.** I'll do this before next week.

**Comment by dankelley** [comment]
It will be in branch `issue_5_new_OO` whilst under development, and then I'll merge it into `main`.  (Note that I have not started a `develop` branch yet.)

**Comment by dankelley** [comment]
A test of ideas:

``` r
rm(list=ls())
library(testthat)
chain <- function(type, buoyancyPerMeter, length, diameter, CD)
{
    rval <- list(type=type, buoyancy=buoyancyPerMeter, length=length, width=width, CD=CD)
    class(rval) <- c("mooring", "chain")
    rval
}
float <- function(type, buoyancy, height, area, CD)
{
    rval <- list(type=type, buoyancy=buoyancy, height=height, area=area, CD=CD)
    class(rval) <- c("mooring", "float")
    rval
}
wire <- function(type, buoyancyPerMeter, length, width, CD)
{
    rval <- list(type=type, buoyancy=buoyancyPerMeter, length=length, width=width, CD=CD)
    class(rval) <- c("mooring", "wire")
    rval
}

mooring <- function(...)
{
    dots <- list(...)
    w <- which(sapply(dots, function(x) !inherits(x, "mooring")))
    if (length(w))
        stop("these are the indices of elements that are not of class 'mooringElement': ", paste(w, collapse=" "))
    rval <- dots
    class(rval) <- "mooring"
    rval
}

# FIXME: why not do discretization within knockdown?  That would be more discoverable.

knockdown <- function(m, u=0)
{
    nclass <- length(class(m))
    if (nclass != 1)
        stop("first argument must be the output of mooring()")
    # do the calculation, storing x,y,phi,T within the object (decide where)
    m
}

F <- float(type="Kiel SFS40in", buoyancy=320, height=1, area=pi, CD=0.65)
W <- wire(type="1/4 wire/jack", buoyancyPerMeter=-0.13, length=10, width=0.008, CD=1.3)
M <- mooring(W, F)
str(M)
#> List of 2
#>  $ :List of 5
#>   ..$ type    : chr "1/4 wire/jack"
#>   ..$ buoyancy: num -0.13
#>   ..$ length  : num 10
#>   ..$ width   : num 0.008
#>   ..$ CD      : num 1.3
#>   ..- attr(*, "class")= chr [1:2] "mooring" "wire"
#>  $ :List of 5
#>   ..$ type    : chr "Kiel SFS40in"
#>   ..$ buoyancy: num 320
#>   ..$ height  : num 1
#>   ..$ area    : num 3.14
#>   ..$ CD      : num 0.65
#>   ..- attr(*, "class")= chr [1:2] "mooring" "float"
#>  - attr(*, "class")= chr "mooring"
str(M[[1]])
#> List of 5
#>  $ type    : chr "1/4 wire/jack"
#>  $ buoyancy: num -0.13
#>  $ length  : num 10
#>  $ width   : num 0.008
#>  $ CD      : num 1.3
#>  - attr(*, "class")= chr [1:2] "mooring" "wire"
expect_error(knockdown(F), "first argument must be the output of mooring()")
K <- knockdown(M)
str(K)
#> List of 2
#>  $ :List of 5
#>   ..$ type    : chr "1/4 wire/jack"
#>   ..$ buoyancy: num -0.13
#>   ..$ length  : num 10
#>   ..$ width   : num 0.008
#>   ..$ CD      : num 1.3
#>   ..- attr(*, "class")= chr [1:2] "mooring" "wire"
#>  $ :List of 5
#>   ..$ type    : chr "Kiel SFS40in"
#>   ..$ buoyancy: num 320
#>   ..$ height  : num 1
#>   ..$ area    : num 3.14
#>   ..$ CD      : num 0.65
#>   ..- attr(*, "class")= chr [1:2] "mooring" "float"
#>  - attr(*, "class")= chr "mooring"
```

<sup>Created on 2021-05-28 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by dankelley** [comment]
This is done in the "develop" branch, commit 1265625ef31e7770d533eccd7a31588e6076f698.

**Comment by dankelley** [comment]
This is merged into "main", commit 8074649043a3c71d29f18cd91bfaf5e5a8cdee82, and so I'm closing the issue.  (There is a compatibility check: if you try using `+`, you'll get an error message).  All docs have been updated.

**Comment by richardsc** [comment]
(psssst. you should update the README example to no longer include `+` 😄 )

**Comment by richardsc** [comment]
Also, some of the help pages have orphaned references:

https://dankelley.github.io/mooring/reference/anchor.html

**Comment by dankelley** [comment]
Both fixed in "main" commit ee96f12ede61a698d236845eb76da21c92836f36. Thanks for the catch(es)!
